src/eos-companion-app-service-managed-cache.c

```diff
@@ -0,0 +1,201 @@
+/* Copyright 2018 Endless Mobile, Inc.
+ *
+ * eos-companion-app-service is free software: you can redistribute it and/or
+ * modify it under the terms of the GNU Lesser General Public License as
+ * published by the Free Software Foundation, either version 2.1 of the
+ * License, or (at your option) any later version.
+ *
+ * eos-companion-app-service is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public
+ * License along with eos-companion-app-service.  If not, see
+ * <http://www.gnu.org/licenses/>.
+ */
+
+#include "eos-companion-app-service-managed-cache.h"
+#include "eos-companion-app-service-managed-cache-private.h"
+
+typedef struct _EosCompanionAppServiceManagedCache {
+  GObject object;
+} EosCompanionAppServiceManagedCache;
+
+typedef struct _EosCompanionAppServiceManagedCachePrivate
+{
+  ThreadsafeRegisteredCache *registered_cache;
+} EosCompanionAppServiceManagedCachePrivate;
+
+G_DEFINE_TYPE_WITH_PRIVATE (EosCompanionAppServiceManagedCache,
+                            eos_companion_app_service_managed_cache,
+                            G_TYPE_OBJECT)
+
+struct _ThreadsafeRegisteredCache {
+  GMutex      mutex;
+  GHashTable *ht;
+};
+
+ThreadsafeRegisteredCache *
+threadsafe_registered_cache_new (GDestroyNotify value_destroy_func)
+{
+  ThreadsafeRegisteredCache *cache = g_new0 (ThreadsafeRegisteredCache, 1);
+
+  cache->ht = g_hash_table_new_full (g_str_hash,
+                                     g_str_equal,
+                                     g_free,
+                                     value_destroy_func);
+
+  return cache;
+}
+
+void
+threadsafe_registered_cache_free (ThreadsafeRegisteredCache *cache)
```

Comment:
It's problematic to have this `free()` function that locks the mutex. What if a thread calls `free()` while another one tries to look something up or add a new value? If this runs first, the hash table will be invalid by the time the other thread gets to it.
Also, `g_hash_table_unref()` is presumably already implemented as an atomic/threadsafe operation?

Reply:
Argh, yeah that will be an issue :(

I'll need to check if GHashTable is already threadsafe, if so we can remove this structure entirely.